### PR TITLE
[wasm] Migrate rest of the JS interop functions to out params

### PR DESCRIPTION
**Pull request by maraf**
A follow up on the https://github.com/dotnet/runtime/pull/65994.

Migrate rest of the JS interop functions to out params instead of return.
- `mono_wasm_invoke_js_with_args`
- `mono_wasm_cancel_promise`
- `mono_wasm_compile_function`

### CODE REVIEW

**Review by pavelsavara — APPROVED** [review]

**Comment by pavelsavara** [comment]
@kg please have look as well.

**Comment by kg** [comment]
Any icalls we change the signature of like these need to have their name changed, because any third party callers (like blazor, uno etc) will crash or corrupt the heap by passing the wrong number of arguments (or reading a return value that isn't there). The convention I selected is a 'Ref' suffix at the end of the name for C#/C methods, and for JS methods either 'ref' (if they take a MonoObjectRef) or 'root' (if they take a WasmRoot)

File: src/mono/wasm/runtime/method-calls.ts [hunk 1]
```diff
@@ -301,17 +301,21 @@ export function mono_call_assembly_entry_point(assembly: string, args?: any[], s
     return mono_bind_assembly_entry_point(assembly, signature)(...args);
 }
 
-export function mono_wasm_invoke_js_with_args(js_handle: JSHandle, method_name: MonoString, args: MonoArray, is_exception: Int32Ptr): any {
-    const argsRoot = mono_wasm_new_root(args), nameRoot = mono_wasm_new_root(method_name);
+export function mono_wasm_invoke_js_with_args(js_handle: JSHandle, method_name: MonoString, args: MonoArray, is_exception: Int32Ptr, result_address: MonoObjectRef): any {
```
Comment (kg) [inline src/mono/wasm/runtime/method-calls.ts:304]: method_name should become MonoStringRef, which will turn nameRoot into an external root. args also needs to become MonoObjectRef and argsRoot becomes external root as well

File: src/mono/wasm/runtime/method-calls.ts [hunk 2]
```diff
@@ -617,9 +620,13 @@ export function mono_wasm_invoke_js(code: MonoString, is_exception: Int32Ptr): M
 // Compiles a JavaScript function from the function data passed.
 // Note: code snippet is not a function definition. Instead it must create and return a function instance.
 // code like `return function() { App.call_test_method(); };`
-export function mono_wasm_compile_function(code: MonoString, is_exception: Int32Ptr): MonoObject {
-    if (code === MonoStringNull)
-        return MonoStringNull;
+export function mono_wasm_compile_function(code: MonoString, is_exception: Int32Ptr, result_address: MonoObjectRef): void {
```
Comment (kg) [inline src/mono/wasm/runtime/method-calls.ts:623]: code needs to be MonoStringRef

**Review by kg — CHANGES_REQUESTED** [review]
Thanks for doing the work of spotting these issues and updating the code!

**Comment by maraf** [comment]
> Any icalls we change the signature of like these need to have their name changed, because any third party callers (like blazor, uno etc) will crash or corrupt the heap by passing the wrong number of arguments (or reading a return value that isn't there). The convention I selected is a 'Ref' suffix at the end of the name for C#/C methods, and for JS methods either 'ref' (if they take a MonoObjectRef) or 'root' (if they take a WasmRoot)

Thanks! Now I see it in your PR, I should have read it before.

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.JS.Owned.cs [hunk 3]
```diff
@@ -81,6 +84,7 @@ public static void GetTaskSourceTaskRef(int tcsGCHandle, out object result)
         {
             GCHandle handle = (GCHandle)(IntPtr)tcsGCHandle;
             // this is JS owned Normal handle. We always have a Target
+            Debug.Assert(handle.Target != null, "TaskCompletionSource should not be null.");
```
Comment (kg) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.JS.Owned.cs:87]: Will this assert run (and throw) in Release builds? If not we'll get a NullReferenceException at the end, so we should just turn it into an if -> throw.

**Review by kg — APPROVED** [review]

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.JS.Owned.cs (hunk 3 above)
Comment (maraf) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.JS.Owned.cs:87]: Sorry, this shouldn't be here. Thanks for the catch.

**Pull request merged**